phpseclib/Crypt/Random.php

```diff
@@ -260,6 +260,44 @@ function crypt_random_string($length)
     }
 }
 
+if (!function_exists('phpseclib_safe_serialize')) {
+    /**
+     * Safely serialize variables
+     *
+     * If a class has a private __sleep() method it'll give a fatal error on PHP 5.2 and earlier.
+     * PHP 5.3 will emit a warning.
+     *
+     * @param mixed $arr
+     * @param array $refs optional
+     * @access public
+     */
+    function phpseclib_safe_serialize($arr, $refs = array())
+    {
+        if (is_object($arr)) {
+            return '';
+        }
+        if (!is_array($arr)) {
+            return serialize($arr);
+        }
+        $safearr = array();
+        foreach (array_keys($arr) as $key) {
+            if (is_object($arr[$key])) {
+                continue;
+            }
+            // without this recursive arrays (eg. $a = []; $a[] = &$a;) would result in an
+            // "Allowed memory size of ... bytes exhausted" Fatal error
+            foreach ($refs as $ref) {
+                if ($ref === $arr[$key]) {
```

Comment:
Thanks!

> I don't think `$key !== '__phpseclib_marker' &&` is necessary

* Without it, each '__phpseclib_marker' key is also serialized. E.g., the test variables in https://github.com/phpseclib/phpseclib/pull/934/files#r52842344 above result in the following output (on PHP 7.0.3):

  ```
  a:9:{s:4:"_GET";s:36:"a:1:{s:18:"__phpseclib_marker";b:1;}";s:5:"_POST";s:36:"a:1:
  {s:18:"__phpseclib_marker";b:1;}";s:7:"_COOKIE";s:36:"a:1:
  {s:18:"__phpseclib_marker";b:1;}";s:6:"_FILES";s:36:"a:1:
  {s:18:"__phpseclib_marker";b:1;}";s:1:"a";s:60:"a:2:
  {s:4:"key1";s:6:"value1";s:18:"__phpseclib_marker";b:1;}";s:1:"b";s:108:"a:2:{i:0;s:60:"a:2:
  {s:4:"key1";s:6:"value1";s:18:"__phpseclib_marker";b:1;}";s:18:"__phpseclib_marker";b:1;}
  ";s:1:"c";s:157:"a:2:{i:0;s:108:"a:2:{i:0;s:60:"a:2:
  {s:4:"key1";s:6:"value1";s:18:"__phpseclib_marker";b:1;}";s:18:"__phpseclib_marker";b:1;}
  ";s:18:"__phpseclib_marker";b:1;}";s:1:"z";s:84:"a:2:{i:0;s:36:"a:1:
  {s:18:"__phpseclib_marker";b:1;}";s:18:"__phpseclib_marker";b:1;}";s:18:"__phpseclib_marker";b:1;}
  ```

  As opposed to the following otherwise:

  ```
  a:8:{s:4:"_GET";s:6:"a:0:{}";s:5:"_POST";s:6:"a:0:{}";s:7:"_COOKIE";s:6:"a:0:{}";s:6:"_FILES";s:6:"a:0:
  {}";s:1:"a";s:30:"a:1:{s:4:"key1";s:6:"value1";}";s:1:"b";s:48:"a:1:{i:0;s:30:"a:1:
  {s:4:"key1";s:6:"value1";}";}";s:1:"c";s:66:"a:1:{i:0;s:48:"a:1:{i:0;s:30:"a:1:
  {s:4:"key1";s:6:"value1";}";}";}";s:1:"z";s:23:"a:1:{i:0;s:6:"a:0:{}";}";}
  ```

  So, roughly the double RAM consumption on that non-representative test. (And a similar proportion on PHP 4.4.9.) I don't know how much impact that has in a typical PHP environment though, with respect to the number of existing global variables containing arrays (don't need to be circular).

* Also, the comment on the following line would need to be removed then as well: https://github.com/terrafrost/phpseclib/pull/6/files#diff-b08ee4e00815fdb38838fb5016bd2ee4R285.